### PR TITLE
feat: Kubernetes Auth Method

### PR DESCRIPTION
**Pull request by tysoekong**
### Summary

Properties and other instructions in relevant to the change proposed in "Kong Enterprise PR 3824".

The feature being added is two parts:

1. Kubernetes Auth Method compatibility when using Kong Keyring
2. Kubernetes Auth Method compatibility when using the Kong Vault Backend "HCV" (Hashicorp Vault)

### Reason

To add the Kubernetes Auth support that some customers require for Hashicorp Vault connections across two different Kong features.

### Testing

We will be able to view the changes on Netlify - @tysoekong will be able to confirm that all shell outputs, as in the examples, still correspond to their appearance in the docs preview. @tysoekong will update the examples accordingly.

### CODE REVIEW

File: src/gateway/reference/configuration.md [hunk 1]
```diff
@@ -3595,6 +3595,24 @@ Defines the names of the Vault v2 KV path at which symmetric keys are found.
 
 
 
+### keyring_vault_auth_method
```
Comment (Guaris) [inline src/gateway/reference/configuration.md:3598]: if you haven't done it already, this doc is generated by editing this file in `kong-ee` https://github.com/Kong/kong-ee/blob/master/kong.conf.default

File: src/gateway/kong-enterprise/db-encryption.md [hunk 2]
```diff
@@ -385,7 +385,7 @@ objects. For information on which plugin fields are encrypted, see
 
 Kong's keyring mechanism can integrate directly with [HashiCorp Vault](https://www.vaultproject.io/) for keyring storage and versioning. In this model, Kong nodes read keyring material directly from a Vault KV secrets engine, rather than generating and disseminating keyring material around the cluster.
 
-To configure Kong to use Vault for keyring storage, set the `keyring_strategy` configuration value to `vault`. Leveraging Vault also requires defining a host, mount point, and token for Vault access. See the Kong configuration reference for more details.
+To configure Kong to use Vault for keyring storage, set the `keyring_strategy` configuration value to `vault`. Leveraging Vault also requires defining a host, mount point, and token or [Kubernetes service account role](https://developer.hashicorp.com/vault/docs/auth/kubernetes) for Vault access. See the Kong configuration reference for more details.
```
Comment (Guaris) [inline src/gateway/kong-enterprise/db-encryption.md:388]:
```suggestion
To configure Kong to use Vault for keyring storage, set the `keyring_strategy` configuration value to `vault`. Leveraging Vault also requires defining a host, mount point, and token or [Kubernetes service account role](https://developer.hashicorp.com/vault/docs/auth/kubernetes) for Vault access. See the [configuration reference](/gateway/latest/reference/configuration/#keyring_vault_kube_api_token_file) for more details.
```

File: src/gateway/kong-enterprise/db-encryption.md [hunk 3]
```diff
@@ -404,7 +404,12 @@ To provide a new key, add a new version to the Vault secrets engine at the confi
 
 ### Vault Permissions
 
-In order to communicate with Vault, Kong must be provided a Vault token for access. The token must be associated with a policy that allows the `read` and `list` actions on the path where keyring secrets are stored. Kong does not write keyring material to the Vault cluster.
+In order to communicate with Vault, Kong must be provided with either:
```
Comment (Guaris) [inline src/gateway/kong-enterprise/db-encryption.md:407]:
```suggestion
In order to communicate with Vault, {{site.base_gateway}} must be provided with either:
```

File: src/gateway/kong-enterprise/db-encryption.md [hunk 4]
```diff
@@ -404,7 +404,12 @@ To provide a new key, add a new version to the Vault secrets engine at the confi
...
+
+* A Vault token
+* A Vault role granted to the Kubernetes service-account of the running pod
+
+The token or Kubernetes auth role must be associated with a policy that allows the `read` and `list` actions on the path where keyring secrets are stored. Kong does not write keyring material to the Vault cluster.
```
Comment (Guaris) [inline src/gateway/kong-enterprise/db-encryption.md:412]:
```suggestion
* A Vault token.
* The Vault role applied to the Kubernetes service account of the running pod.

The token or Kubernetes auth role must be associated with a policy that allows the `read` and `list` actions on the path where keyring secrets are stored. Kong does not write keyring material to the Vault cluster.
```

File: src/gateway/kong-enterprise/secrets-management/backends/hashicorp-vault.md [hunk 5]
```diff
@@ -9,15 +9,31 @@ badge: enterprise
 
 ## Environment variables
 
+Using static Vault token authentication:
```
Comment (Guaris) [inline src/gateway/kong-enterprise/secrets-management/backends/hashicorp-vault.md:12]:
```suggestion
Static Vault token authentication:
```

File: src/gateway/kong-enterprise/secrets-management/backends/hashicorp-vault.md [hunk 6]
```diff
@@ -9,15 +9,31 @@ badge: enterprise
...
 export KONG_VAULT_HCV_PROTOCOL=<protocol(http|https)>
 export KONG_VAULT_HCV_HOST=<hostname>
 export KONG_VAULT_HCV_PORT=<portnumber>
 export KONG_VAULT_HCV_MOUNT=<mountpoint>
 export KONG_VAULT_HCV_KV=<v1|v2>
+export KONG_VAULT_HCV_AUTH_METHOD=token
 export KONG_VAULT_HCV_TOKEN=<tokenstring>
 ```
 
+Or, using Kubernetes service-account role authentication:
```
Comment (Guaris) [inline src/gateway/kong-enterprise/secrets-management/backends/hashicorp-vault.md:25]:
```suggestion
Kubernetes service account role authentication:
```

File: src/gateway/reference/configuration.md [hunk 7]
```diff
@@ -3604,6 +3622,28 @@ Defines the token value used to communicate with the v2 KV Vault HTTP(S) API.
 
 
 
+### keyring_vault_kube_role
+{:.badge .enterprise}
+
+Defines the Hashicorp Vault role that will be assumed using the Kubernetes service
```
Comment (Guaris) [inline src/gateway/reference/configuration.md:3628]:
```suggestion
Defines the HashiCorp Vault role for the Kubernetes service
```

File: src/gateway/reference/configuration.md [hunk 8]
```diff
@@ -3604,6 +3622,28 @@ Defines the token value used to communicate with the v2 KV Vault HTTP(S) API.
...
+`keyring_vault_auth_method` must be set to `kubernetes` for this to activate.
+
+**Default:** `default`
+
+
+
+### keyring_vault_kube_api_token_file
+{:.badge .enterprise}
+
+Defines where the Kubernetes service account token should be read from the pod's
```
Comment (Guaris) [inline src/gateway/reference/configuration.md:3640]:
```suggestion
Defines where the Kubernetes service account token should be read from on the pod's
```

**Review by Guaris — CHANGES_REQUESTED** [review]

File: src/gateway/reference/configuration.md (hunk 1 above)
Comment (tysoekong) [reply src/gateway/reference/configuration.md:3598]: I have now done that.

**Comment by acgoldsmith** [comment]
I'm adding the DNM tag while we're waiting to confirm the feature is releasing in 3.1 - watching for approval form Anand on https://konghq.atlassian.net/browse/FTI-4231

**Review by Guaris — APPROVED** [review]

**Comment by acgoldsmith** [comment]
Confirmed on ticket that is included in 3.1 scope

**Pull request merged**